executor/cluster_reader_test.go

```diff
@@ -0,0 +1,235 @@
+	s.store = store
+	s.dom = dom
+}
+
+func (s *testClusterReaderSuite) TearDownSuite(c *C) {
+	s.dom.Close()
+	s.store.Close()
+}
+
+func (s *testClusterReaderSuite) TestTiDBClusterConfig(c *C) {
+	// mock PD http server
+	router := mux.NewRouter()
+
+	type mockServer struct {
+		address string
+		server  *httptest.Server
+	}
+	const testServerCount = 3
+	var testServers []*mockServer
+	for i := 0; i < testServerCount; i++ {
+		server := httptest.NewServer(router)
+		address := strings.TrimPrefix(server.URL, "http://")
+		testServers = append(testServers, &mockServer{
+			address: address,
+			server:  server,
+		})
+	}
+	defer func() {
+		for _, server := range testServers {
+			server.server.Close()
+		}
+	}()
+
+	// We check the counter to valid how many times request has been sent
+	var requestCounter int32
+	var mockConfig = func() (map[string]interface{}, error) {
+		atomic.AddInt32(&requestCounter, 1)
+		configuration := map[string]interface{}{
+			"key1": "value1",
+			"key2": map[string]string{
+				"nest1": "n-value1",
+				"nest2": "n-value2",
+			},
+		}
+		return configuration, nil
+	}
+
+	// pd config
+	router.Handle(pdapi.Config, fn.Wrap(mockConfig))
+	// TiDB/TiKV config
+	router.Handle("/config", fn.Wrap(mockConfig))
+
+	// mock servers
+	servers := []string{}
+	for _, typ := range []string{"tidb", "tikv", "pd"} {
+		for _, server := range testServers {
+			servers = append(servers, strings.Join([]string{typ, server.address, server.address}, ","))
+		}
+	}
+
+	fpName := "github.com/pingcap/tidb/executor/mockClusterConfigServerInfo"
+	fpExpr := strings.Join(servers, ";")
+	c.Assert(failpoint.Enable(fpName, fmt.Sprintf(`return("%s")`, fpExpr)), IsNil)
+	defer func() { c.Assert(failpoint.Disable(fpName), IsNil) }()
+
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustQuery("select type, `key`, value from information_schema.cluster_config").Check(testkit.Rows(
+		"tidb key1 value1",
+		"tidb key2.nest1 n-value1",
+		"tidb key2.nest2 n-value2",
+		"tidb key1 value1",
+		"tidb key2.nest1 n-value1",
+		"tidb key2.nest2 n-value2",
+		"tidb key1 value1",
+		"tidb key2.nest1 n-value1",
+		"tidb key2.nest2 n-value2",
+		"tikv key1 value1",
+		"tikv key2.nest1 n-value1",
+		"tikv key2.nest2 n-value2",
+		"tikv key1 value1",
+		"tikv key2.nest1 n-value1",
+		"tikv key2.nest2 n-value2",
+		"tikv key1 value1",
+		"tikv key2.nest1 n-value1",
+		"tikv key2.nest2 n-value2",
+		"pd key1 value1",
+		"pd key2.nest1 n-value1",
+		"pd key2.nest2 n-value2",
+		"pd key1 value1",
+		"pd key2.nest1 n-value1",
+		"pd key2.nest2 n-value2",
+		"pd key1 value1",
+		"pd key2.nest1 n-value1",
+		"pd key2.nest2 n-value2",
+	))
+	warnings := tk.Se.GetSessionVars().StmtCtx.GetWarnings()
+	c.Assert(len(warnings), Equals, 0, Commentf("unexpected warnigns: %+v", warnings))
+	c.Assert(requestCounter, Equals, int32(9))
+
+	var cases = []struct {
+		sql      string
+		reqCount int32
+	}{
+		{
+			sql:      "select * from information_schema.cluster_config",
+			reqCount: 9,
```

Comment:
Please check the row count of the result.

Reply:
addressed, thx.